sdktests/server_side_hooks.go

```diff
@@ -0,0 +1,307 @@
+	t.Run("data propagates from before to after for migrations", beforeEvaluationDataPropagatesToAfterMigration)
+}
+
+func executesBeforeEvaluationStage(t *ldtest.T) {
+	t.Run("without detail", func(t *ldtest.T) { executesBeforeEvaluationStageDetail(t, false) })
+	t.Run("with detail", func(t *ldtest.T) { executesBeforeEvaluationStageDetail(t, true) })
+	t.Run("for migrations", executesBeforeEvaluationStageMigration)
+}
+
+func executesAfterEvaluationStage(t *ldtest.T) {
+	t.Run("without detail", func(t *ldtest.T) { executesAfterEvaluationStageDetail(t, false) })
+	t.Run("with detail", func(t *ldtest.T) { executesAfterEvaluationStageDetail(t, true) })
+	t.Run("for migrations", executesAfterEvaluationStageMigration)
+}
+
+func beforeEvaluationDataPropagatesToAfter(t *ldtest.T) {
+	t.Run("without detail", func(t *ldtest.T) { beforeEvaluationDataPropagatesToAfterDetail(t, false) })
+	t.Run("with detail", func(t *ldtest.T) { beforeEvaluationDataPropagatesToAfterDetail(t, true) })
+}
+
+type VariationParameters struct {
+	name         string
+	flagKey      string
+	defaultValue ldvalue.Value
+	valueType    servicedef.ValueType
+	detail       bool
+}
+
+func variationTestParams(detail bool) []VariationParameters {
+	return []VariationParameters{{
+		name:         "for boolean variation",
+		flagKey:      "bool-flag",
+		defaultValue: ldvalue.Bool(false),
+		valueType:    servicedef.ValueTypeBool,
+		detail:       detail,
+	},
+		{
+			name:         "for string variation",
+			flagKey:      "string-flag",
+			defaultValue: ldvalue.String("default"),
+			valueType:    servicedef.ValueTypeString,
+			detail:       detail,
+		},
+		{
+			name:         "for double variation",
+			flagKey:      "number-flag",
+			defaultValue: ldvalue.Float64(3.14),
+			valueType:    servicedef.ValueTypeDouble,
+			detail:       detail,
+		},
+		{
+			name:         "for int variation",
+			flagKey:      "number-flag",
+			defaultValue: ldvalue.Int(0xDEADBEEF),
+			valueType:    servicedef.ValueTypeInt,
+			detail:       detail,
+		},
+		{
+			name:         "for json variation",
+			flagKey:      "json-flag",
+			defaultValue: ldvalue.ObjectBuild().Build(),
+			valueType:    servicedef.ValueTypeInt,
+			detail:       detail,
+		},
+	}
+}
+
+func executesBeforeEvaluationStageDetail(t *ldtest.T, detail bool) {
+	testParams := variationTestParams(detail)
+
+	hookName := "executesBeforeEvaluationStage"
+	client, hooks := createClientForHooks(t, []string{hookName}, nil)
+	defer hooks.Close()
+
+	for _, testParam := range testParams {
+		t.Run(testParam.name, func(t *ldtest.T) {
+			client.EvaluateFlag(t, servicedef.EvaluateFlagParams{
+				FlagKey:      testParam.flagKey,
+				Context:      o.Some(ldcontext.New("user-key")),
+				ValueType:    testParam.valueType,
+				DefaultValue: testParam.defaultValue,
+			})
+
+			hooks.ExpectCall(t, hookName, 1*time.Second, func(payload servicedef.HookExecutionPayload) bool {
```

Comment:
This 1 second should be pulled out into a constant (or perhaps use one of the existing timeout constants?)